runtime/src/stakes.rs

```diff
@@ -235,18 +235,9 @@ impl Stakes<StakeAccount> {
                 None => return Err(Error::VoteAccountNotFound(*pubkey)),
                 Some(account) => account,
             };
-            // Ignoring rent_epoch until the feature for
-            // preserve_rent_epoch_for_rent_exempt_accounts is activated.
             let vote_account = vote_account.account();
-            if vote_account.lamports() != account.lamports()
-                || vote_account.owner() != account.owner()
-                || vote_account.executable() != account.executable()
-                || vote_account.data() != account.data()
-            {
-                error!(
-                    "vote account mismatch: {}, {:?}, {:?}",
-                    pubkey, vote_account, account
-                );
+            if vote_account != &account {
+                error!("vote account mismatch: {pubkey}, {vote_account:?}, {account:?}");
                 return Err(Error::VoteAccountMismatch(*pubkey));
```

Comment:
Does this halt the validator? IIRC, using the cache is not enabled yet on mnb?

Reply:
> Does this halt the validator? 

It _does_ halt the validator _at start up_.

This code is only invoked at startup when deserializing bank:
https://github.com/solana-labs/solana/blob/0be194145/runtime/src/bank.rs#L1828-L1841
in order to obtain `Stakes<StakeAccount>` from `Stakes<Delegation>` for backward compatibility reasons mentioned in the link. As part of that, it performs these sanity checks.

> IIRC, using the cache is not enabled yet on mnb?

We are not using stakes-cache for rewards calculations (yet), but stakes-cache is already used so many other places, including epoch-stakes calculations. Pubkeys for stakes accounts used in rewards calculations are also already obtained from stakes-cache.

Reply:
as a side note, I wonder why we aren't building this cache on startup instead of serializing it. We scan all accounts to generate the index. It seems we could recreate this then. Bank deserialization at load does take a while and all of this is duplicate information derivable from append vecs at startup. Then we would have no inconsistency checks and it seems we could remove code.

Reply:
Not sure where to obtain correct `stake_history: StakeHistory` field in stakes-cache, but, yeah, as for stake/vote accounts, those can be obtained from accounts-db. It would need a full accounts-db scan though.

I think since we already do serialize it, it would be just easier/faster to go with it.

> Then we would have no inconsistency checks and it seems we could remove code.

Any inconsistency spotted by sanity checks here is only a symptom that there is a bug somewhere. The actual culprit of that inconsistency is somewhere else.